src/databricks/labs/remorph/reconcile/connectors/data_source.py

```diff
@@ -1,60 +1,46 @@
+from pyspark.sql import DataFrame
 
+from databricks.labs.remorph.reconcile.exception import MockDataNotAvailableException
 from databricks.labs.remorph.reconcile.recon_config import JdbcReaderOptions, Schema
-from databricks.sdk import WorkspaceClient
 
 
 class DataSource(ABC):
-    # TODO need to remove connection_params
-    def __init__(
-        self,
-        engine: str,
-        spark: SparkSession,
-        ws: WorkspaceClient,
-        scope: str,
-    ):
-        self.engine = engine
-        self.spark = spark
-        self.ws = ws
-        self.scope = scope
 
     @abstractmethod
-    def read_data(self, catalog: str, schema: str, query: str, options: JdbcReaderOptions) -> DataFrame:
+    def read_query_data(
+        self, catalog: str, schema: str, table: str, query: str, options: JdbcReaderOptions | None
+    ) -> DataFrame:
         return NotImplemented
 
     @abstractmethod
     def get_schema(self, catalog: str, schema: str, table: str) -> list[Schema]:
         return NotImplemented
 
-    def _get_jdbc_reader(self, query, jdbc_url, driver):
-        return (
-            self.spark.read.format("jdbc")
-            .option("url", jdbc_url)
-            .option("driver", driver)
-            .option("dbtable", f"({query}) tmp")
-        )
-
-    @staticmethod
-    def _get_jdbc_reader_options(options: JdbcReaderOptions):
-        return {
-            "numPartitions": options.number_partitions,
-            "partitionColumn": options.partition_column,
-            "lowerBound": options.lower_bound,
-            "upperBound": options.upper_bound,
-            "fetchsize": options.fetch_size,
-        }
-
-    def _get_secrets(self, key_name: str):
-        key = self.engine + '_' + key_name
-        return self.ws.secrets.get_secret(self.scope, key)
-
-    @staticmethod
-    def _get_table_or_query(catalog: str, schema: str, query: str) -> str:
-        if re.search('select', query, re.IGNORECASE):
-            return query.format(catalog_name=catalog, schema_name=schema)
-        if catalog and catalog != "hive_metastore":
-            return f"select * from {catalog}.{schema}.{query}"
-        return f"select * from {schema}.{query}"
+
+class MockDataSource(DataSource):
+
+    def __init__(
+        self,
+        dataframe_repository: dict[(str, str, str), DataFrame],
+        schema_repository: dict[(str, str, str), list[Schema]],
+    ):
+        self._dataframe_repository: dict[(str, str, str), DataFrame] = dataframe_repository
+        self._schema_repository: dict[(str, str, str), list[Schema]] = schema_repository
+
+    def read_query_data(
+        self, catalog: str, schema: str, table: str, query: str, options: JdbcReaderOptions | None
+    ) -> DataFrame:
+        mock_df = self._dataframe_repository.get((catalog, schema, query))
+        if not mock_df:
+            raise MockDataNotAvailableException(f"data is not mocked for the combination : {(catalog, schema, query)}")
```

Comment:
are you going to return the same exception when data is not returned in real scenario?

Reply:
refactored